compress/lz4/lz4.go

```diff
@@ -54,55 +42,107 @@ func (c *Codec) CompressionCodec() format.CompressionCodec {
 }
 
 func (c *Codec) NewReader(r io.Reader) (compress.Reader, error) {
-	lzr := lz4.NewReader(r)
-	err := lzr.Apply(
-		lz4.ConcurrencyOption(c.concurrency()),
-	)
-	if err != nil {
-		return nil, err
-	}
-	return reader{lzr}, nil
+	return &reader{reader: r}, nil
 }
 
 func (c *Codec) NewWriter(w io.Writer) (compress.Writer, error) {
-	lzw := lz4.NewWriter(w)
-	err := lzw.Apply(
-		lz4.BlockChecksumOption(false),
-		lz4.ChecksumOption(false),
-		lz4.BlockSizeOption(c.blockSize()),
-		lz4.CompressionLevelOption(c.level()),
-		lz4.ConcurrencyOption(c.concurrency()),
-	)
-	if err != nil {
-		return nil, err
-	}
-	return writer{lzw}, nil
+	return &writer{writer: w, compressor: lz4.CompressorHC{Level: c.Level}}, nil
 }
 
-func (c *Codec) concurrency() int {
-	if c.Concurrency != 0 {
-		return c.Concurrency
-	}
-	return DefaultConcurrency
+type reader struct {
+	buffer bytes.Buffer
+	data   []byte
+	offset int
+	reader io.Reader
+}
+
+func (r *reader) Close() error {
+	r.offset = len(r.data)
+	r.reader = nil
+	return nil
 }
 
-func (c *Codec) blockSize() BlockSize {
-	if c.BlockSize != 0 {
-		return c.BlockSize
+func (r *reader) Reset(rr io.Reader) error {
+	r.buffer.Reset()
+	r.data = r.data[:0]
+	r.offset = 0
+	r.reader = rr
+	return nil
+}
+
+func (r *reader) Read(b []byte) (n int, err error) {
+	if r.offset == 0 && len(r.data) == 0 {
+		if err := r.decompress(); err != nil {
+			return 0, err
+		}
+	}
+	n = copy(b, r.data[r.offset:])
+	r.offset += n
+	if r.offset == len(r.data) {
+		err = io.EOF
 	}
-	return DefaultBlockSize
+	return n, err
 }
 
-func (c *Codec) level() Level {
-	// zero == Fast
-	return c.Level
+func (r *reader) decompress() error {
+	if r.reader == nil {
+		return io.EOF
+	}
+
+	_, err := r.buffer.ReadFrom(r.reader)
+	if err != nil {
+		return err
+	}
+
+	if size := 3 * r.buffer.Len(); cap(r.data) < size {
+		r.data = make([]byte, size)
+	} else {
+		r.data = r.data[:cap(r.data)]
+	}
+
+	for {
+		n, err := lz4.UncompressBlock(r.buffer.Bytes(), r.data)
+		if err != nil {
```

Comment:
Check for specific error? This feels fragile

Reply:
I looked into that, but the `lz4` package doesn't expose the error values (they're in an `internal/lz4errors` package). Based on what I read of the implementation, the only condition where this function errors is if the output buffer is too short.